jgrapht-core/src/main/java/org/jgrapht/alg/cycle/ChinesePostman.java

```diff
@@ -0,0 +1,302 @@
+ *
+ * This program and the accompanying materials are dual-licensed under
+ * either
+ *
+ * (a) the terms of the GNU Lesser General Public License version 2.1
+ * as published by the Free Software Foundation, or (at your option) any
+ * later version.
+ *
+ * or (per the licensee's choosing)
+ *
+ * (b) the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation.
+ */
+package org.jgrapht.alg.cycle;
+
+import org.jgrapht.Graph;
+import org.jgrapht.GraphPath;
+import org.jgrapht.GraphTests;
+import org.jgrapht.Graphs;
+import org.jgrapht.alg.interfaces.EulerianCycleAlgorithm;
+import org.jgrapht.alg.interfaces.MatchingAlgorithm;
+import org.jgrapht.alg.interfaces.ShortestPathAlgorithm;
+import org.jgrapht.alg.matching.KuhnMunkresMinimalWeightBipartitePerfectMatching;
+import org.jgrapht.alg.matching.blossom.v5.*;
+import org.jgrapht.alg.shortestpath.DijkstraShortestPath;
+import org.jgrapht.alg.util.Pair;
+import org.jgrapht.alg.util.UnorderedPair;
+import org.jgrapht.graph.*;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * This class solves the Chinese Postman Problem (CPP), also known as the Route Inspection Problem.
+ * The CPP asks to find a shortest closed path or circuit that visits every edge of a graph. In
+ * weighted graphs, the circuit of minimal total weight is returned; in unweighted graphs, a circuit
+ * of minimum total length (total number of edges) is returned.
+ * <p>
+ * The algorithm works with directed and undirected graphs which may contain loops and/or multiple
+ * edges. The runtime complexity is O(N^3) where N is the number of vertices in the graph. Mixed
+ * graphs are currently not supported, as solving the CPP for mixed graphs is NP-hard. The graph on
+ * which this algorithm is invoked must be strongly connected; invoking this algorithm on a graph
+ * which is not strongly connected may result in undefined behavior. In case of weighted graphs, all
+ * edge weights must be positive.
+ *
+ * If the input graph is Eulerian (see {@link GraphTests#isEulerian(Graph)} for details) use
+ * {@link HierholzerEulerianCycle} instead.
+ * <p>
+ * The implementation is based on the following paper:<br>
+ * Edmonds, J., Johnson, E.L. Matching, Euler tours and the Chinese postman, Mathematical
+ * Programming (1973) 5: 88. doi:10.1007/BF01580113<br>
+ *
+ * More concise descriptions of the algorithms can be found here:
+ * <ul>
+ * <li><a href=
+ * "http://www.suffolkmaths.co.uk/pages/Maths%20Projects/Projects/Topology%20and%20Graph%20Theory/Chinese%20Postman%20Problem.pdf">CPP
+ * for Undirected graphs</a>
+ * <li><a href=
+ * "https://www-m9.ma.tum.de/graph-algorithms/directed-chinese-postman/index_en.html">CPP for
+ * Directed graphs</a>
+ * </ul>
+ * 
+ * @param <V> the graph vertex type
+ * @param <E> the graph edge type
+ *
+ * @author Joris Kinable
+ * @since September 2018
+ */
+public class ChinesePostman<V, E>
+    implements
+    EulerianCycleAlgorithm<V, E>
```

Comment:
The javadoc for EulerianCycleAlgorithm says "Computes an Eulerian cycle of an Eulerian graph."  But ChinesePostman is intended for use with non-Eulerian graphs,  and necessarily the result is not a Eulerian cycle.  So extending the EulerianCycleAlgorithm interface doesn't seem right.  Maybe we need a generalized superinterface for EulerianCycleAlgorithm?  Mathworld defines a unicursal circuit as "A circuit in which an entire graph is traversed in one route."  So maybe MinimumUnicursalCycleAlgorithm?